### PR TITLE
Use pandas DataFrame to store layer features

### LINKED ISSUE #3778: Use pandas DataFrame to store Layer properties
## 🚀 Feature

I created this mostly for tracking purposes.

The proposal is described in detail in the following document: https://hackmd.io/emsLO3cUR3O9nxSc6YYwUA

## Additional context

Part of #2866

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
This adds pandas as a dependency for napari and uses `pandas.DataFrame` to store a new `features` property that replaces the existing `properties` and `property_choices` properties in the relevant layers.

For more context and motivation around this change, see the following doc: https://hackmd.io/emsLO3cUR3O9nxSc6YYwUA

Despite the size of the PR, the approach is fairly simple and doesn't require too many changes to make basic usage work. I added a little bit of refactoring to simplify usage and make it more consistent across layers.

Some things this PR does not do, mostly to keep it from growing too large.
1. Deprecate `properties` and `property_choices`.
2. Use `features` in all places instead of `properties`, only in places that need it.
3. Replace `current_properties` with a similar default features.

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

This breaks a few public behaviors.

1. Coercion of `properties` and `property_choices` associated with `prepare_properties` by replacing with something much simpler. I think this is somewhat unavoidable when using `pandas.CategoricalDType` to absorb `property_choices`, but it would be good to understand what real use cases this change would break to see if we can fix those.

2. In-place modification of `Layer.properties` (specifically when the DataFrame stores non-numpy arrays/dtypes), though modification of `Layer.features` works as an alternative that should work in all cases.

It also changes the return value of `Layer._get_state` for layers with properties/features. This is non-public, but it looks like it might be exposed to plugins (e.g. see the built-in `napari_write_points`).

# References
Closes #3778
Part of #2866 

# How has this been tested?
- [x] new tests pass
- [x] all existing tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by sofroniewn** [comment]
> https://hackmd.io/emsLO3cUR3O9nxSc6YYwUA

@andy-sweet this doc is great, thanks!!

So if we went with this approach am I right in thinking we could merge this PR as backwards compatible and then deprecate and eventually remove layer.properties? Then we'd just have `layer.features` as a dataframe. That's quite nice, if we're sure we want that!

> However, I'm not convinced by this approach yet mostly because of how it would handle multi-dimensional features, but I thought it was worth getting some feedback now.

As you say here, I think this is the crux of it. I think if we know we're fine with just 2D objects than adding pandas as a dependency and supporting a 2D feature representation feels right. There was maybe some initial hesitancy to committing to pandas (see my historical comments here https://github.com/napari/napari/issues/858#issuecomment-574347608), but now that we're further along I think if we're sure about the 2D stuff then we should go with it, otherwise I feel we'll just end up duplicating the package more and more.

Now for the multidimensional stuff #2917 and #2592 have a lot of great discussion in them, and it seems like we generally want to support them. The question is are we ok saying for now you have to use dtype, object and do your own custom thing. Or do we want to take the plunge into the xarray or equivalent world and add core support for them soon?

What do @jni @tlambert03 @kevinyamauchi @alisterburt @brisvag think here?

**Comment by brisvag** [comment]
I think it would be a mistake to not be ready for n-dimensional properties from the get-go. They would open up a ton of possibilities, especially for time series data. I think this would go hand in hand with using xarray for our data model, but we don't necessarily have to do everything now. Potentially, we should be able to use xarray as a pandas substitute in the degenerate 2D case, or at least set up this pandas implementation so that the switch is backward compatible.

In [this comment about the use of xarray for our datamodel](https://github.com/napari/napari/issues/906#issuecomment-939266939) (there's more stuff in the comments above), I also mentioned how this would help with properties.

(I haven't read the code of this PR yet, will do soon.)

**Comment by brisvag** [comment]
Here's a small showcase of how you could use xarray as drop-in replacement for pandas:

```python
import pandas as pd
import xarray as xr

class A:
    pass
properties = {'a': [1,2,3], 'b': ['x', 'y', 'x'], 'c': [A(), A(), A()]}
df = pd.DataFrame(properties)
ds = xr.Dataset.from_dataframe(df)
ds['a'].to_numpy()  #  array([1, 2, 3])
```

But now we get nd for free:
```python
ds['nd_prop'] = ['time', 'index'], np.random.rand(100, 3)  # tuple(dim_names, data)
# note that 'index' is the name of the first dim just because we came from a dataframe
ds.sel(index=0, time=3)  # some value
```

**Comment by andy-sweet** [comment]
> xarray as drop-in replacement for pandas:

Yeah, I think this PR wouldn't change too much if we used DataSet instead of DataFrame (at least in the layers), so I'm happy to implement that in another WIP PR to see if anything breaks. My gut feeling is that xarray DataSet is the right thing to use here, but my main concern is around [how it handles missing data](http://xarray.pydata.org/en/stable/user-guide/computation.html#missing-values), which effectively means that it only supports floating point and object dtypes (using `np.nan` and `None` respectively to represent missing values).

By contrast numpy has [`np.ma` for masked arrays](https://numpy.org/doc/stable/reference/maskedarray.generic.html#), which xarray doesn't maintain.

```python
In [60]: ma = np.ma.array([1, 2, 3], mask=[False, True, False], dtype=int)

In [61]: ma
Out[61]: 
masked_array(data=[1, --, 3],
             mask=[False,  True, False],
       fill_value=999999)

In [62]: da = xr.DataArray(ma)

In [63]: da.values
Out[63]: array([ 1., nan,  3.])

In [64]: da.values.dtype
Out[64]: dtype('float64')
```

And pandas has [some experimental (but seemingly mature) support](https://pandas.pydata.org/pandas-docs/stable/user_guide/boolean.html#) for integer, boolean, and string, which xarray also can't support.

```python
In [79]: series = pd.Series([1, np.nan, 3], dtype=pd.Int64Dtype())

In [80]: series
Out[80]: 
0       1
1    <NA>
2       3
dtype: Int64

In [81]: da = xr.DataArray(series)

In [82]: da.values
Out[82]: array([1, <NA>, 3], dtype=object)
```

I think it might be OK to limit ourselves to floats and objects for properties/features data, but this may be a more serious problem for data in general (e.g. a labels layer with missing pixels).

The other problems are shared with pandas: it's a large dependency to bring in and it might have unknown performance issues compared to plain numpy. There is [an xarray issue describing something like an xarray-lite](https://github.com/pydata/xarray/issues/3981), but even if that work goes ahead, I imagine it will take at least a few months to be stable.

**Comment by brisvag** [comment]
That's unfortunate. I think that the best chance to get this in xarray would be to expend the experimental support for `pd.Int64Dtype` and friends from pandas.

> I'm happy to implement that in another WIP PR to see if anything breaks.

This would be great, it would definitely help us avoid some simple gotchas!

**Comment by brisvag** [comment]
Relevant issues on the xarray repo:
- pydata/xarray#1194
- pydata/xarray#3955

**Comment by jni** [comment]
Hi everyone! We discussed this at the napari Pacific/US meeting today. @brisvag although your point that we should go to nD straight away is well taken, we all agreed that pandas is a much more familiar and mature target for the vast majority of the community. One of the issues I have is that xarray for 1D data does *not* reduce to pandas, and imho it should.

Believe me @brisvag, I fully agree that we should aim to support n-dimensional data from the get-go — over at scikit-image I [gave myself the title of Guardian of the Dimensions](https://github.com/scikit-image/scikit-image/pull/1038#issuecomment-47189573) 😂 — but in this case I think that we need to push xarray forward to get it to where we want it to be. This means both xarray-lite for image data, and xarray-heavy (probably better titles out there) for feature data. The key requirement I have for xarray-heavy is that, in the case of 1D datasets, it should conform to the nascent [dataframe API](https://github.com/data-apis/dataframe-api) from the data-APIs group. In other words, caring about n-dimensional data means more than supporting n-dimensional data: it means adopting APIs that make sense not just for high n but also for n=1. I'm not saying that xarray's API doesn't, but rather that the API for the 1D case is *so* established, that we want to make sure we support that API as a first class citizen.

So my proposal for this project is:
- adopt pandas dataframes as the supported data model for now
- state everywhere in the docs that we are not targeting the full pandas API, which is ridiculous, but the [dataframe API](https://github.com/data-apis/dataframe-api)
- [work with](https://github.com/data-apis/dataframe-api/issues/59) the dataframe API group and the xarray team to ensure that we can **match that API** when switching from pandas to xarray to ensure the transition is non-breaking
- switch internally from pandas to xarray to be able to support higher-D data while maintaining **1st-class support** for 1D columnar data and the familiar API that people in the PyData community are used to.

@nclack put it quite nicely at the meeting: on the one hand, pandas has limitations that we don't like, but they are well understood and we can in some cases work around them (e.g. using object columns), though with reduced functionality (e.g. no verification of consistent shape). On the other hand, xarray *also* has some limitations, *and* it is less well understood, certainly by the core team but also by the community at large, so there is an element of unknown risk in going full-bore on xarray for this purpose. So it's an easier path to ease into the new API by going with pandas first, and then using xarray in an *API-preserving way* to get our nD support later.

I presume you *already* have data that needs the nD features support @brisvag, so a delay is annoying at best, but imho we should get the pandas API in and then we can work together to build the xarray-based solution. While we do the pandas stuff we should certainly watch out for things that might tie us down or that wouldn't work with nD, though, @andy-sweet.

Last thought: I'm definitely not experienced with the xarray API, so I could be totally mistaken and it already *does* reduce to pandas in 1D. In that case, though, it's still not a terrible plan to start with pandas (well understood) and move to xarray, it'll just be easier, we just need to add the missing data support and we're done! But I suspect we'll need some work on the API still.

**Comment by brisvag** [comment]
I agree 100% with your conclusions! Maybe I wasn't clear enough, but you condensed my thoughts exactly in this line:

> So it's an easier path to ease into the new API by going with pandas first, and then using xarray in an API-preserving way to get our nD support later.

I think we should definitely move forward with pandas for now, but keep an eye out for places where n-D might get blocked. As you saw from my example above, I don't think a later switch will be too hard to accomplish.

---

To summarize for future reference, it seems that before we can move towards xarray we need it to:
- support `pd.NA`, masked arrays or equivalent features. This might be enabled by pandas' custom dtypes like `pd.Int16Dtype` [adopting the array interface](https://github.com/pydata/xarray/issues/3955#issuecomment-611321868) and xarray consuming it as appropriate.
- behave exactly like a pandas dataframe (maybe even be exposable as such) when 1D. I think this might be already the case, or close to.
- adhere to the `__dataframe__` api

**Comment by andy-sweet** [comment]
Thanks for the detailed summary from yesterday's community meeting! From the discussion, it sounds like I should try to get this PR into a reviewable/mergeable form.

@jni : it sounds like you're suggesting that we would make `def features(self)` return a plain old instance of `pandas.DataFrame`, which does not implement the latest version of [the dataframe protocol](https://github.com/data-apis/dataframe-api) (e.g. it doesn't have methods like `num_rows` and `get_column`).

I think it wouldn't be too difficult to wrap/extend pandas' DataFrame/Series to meet the DataFrame/Column protocols, but my guess is we don't want to do that because those protocols are not finalized (right?) and are not familiar to the broad scientific community (e.g. users want `df['confidence']` not `df.get_column('confidence')`).

If that's the case, I guess we want to communicate to users that they should only use the equivalent attributes of `pandas.DataFrame` like `shape`, columns, `__getitem__`, `__setitem__`, `items`, etc. If we only plan to do that in the docstring and user guide, I would vote to omit the return type annotation. What do you think?

In terms of xarray, I might try to add a *Future compatibility* section to the doc to understand how we might be able to later use some version of xarray `DataSet` instead.

**Comment by sofroniewn** [comment]
> If that's the case, I guess we want to communicate to users that they should only use the equivalent attributes of pandas.DataFrame like shape, columns, __getitem__, __setitem__, items, etc. If we only plan to do that in the docstring and user guide, I would vote to omit the return type annotation. What do you think?

Sounds reasonable to me - though I didn't realize that `df['confidence']` wouldn't be supported anymore. That's always been the major way I've used pandas (if I understand right).

I took think many folks won't read/ appreciate what's going on here, but we might just have to bite that bullet when the time comes.

> In terms of xarray, I might try to add a Future compatibility section to the doc to understand how we might be able to later use some version of xarray DataSet instead.

Sounds good to me!

------

Looks like the test fails are related to #3747 and independent of this PR

@andy-sweet I see this PR is still in draft, is there anything outside of doc strings you still want to do (if even that is still needed)? I took a quick look through code and it was looking good

**Comment by andy-sweet** [comment]
> I see this PR is still in draft, is there anything outside of doc strings you still want to do (if even that is still needed)

There are a still a few things to get this out of draft (other than the docstrings).

1. Fix the failing tests that are related to this change (there are a few still).
2. Add the pandas dependency to setup.cfg.
3. Add more test coverage around some of the helpers methods.
4. Decide if we want to use `features` everywhere instead of `properties` (e.g. in things like `ColorManager`). If we do, make those updates.

We have an arch WG meeting in a couple of days to help make some of those decisions and I also booked some time with @jni the day after to push this through, so should be reviewable next week.

**Comment by andy-sweet** [comment]
> 2\. Add the pandas dependency to setup.cfg.

What's the best way to determine the pandas version that should be pinned?

I looked at what I had in my environment (1.3.3) and then made that the minimum version required. Anything else I should check?

**Comment by sofroniewn** [comment]
> What's the best way to determine the pandas version that should be pinned?
>
> I looked at what I had in my environment (1.3.3) and then made that the minimum version required. Anything else I should check?

1.3.3 is latest - out in September.

In theory we could probably support much an older version for now, all the way back to `1.0.0` I suspect (released Jan 29 2020), and then update later as needed. What does @jni recommend here. I'm also fine with 1.3.3 I should add

**Comment by sofroniewn** [comment]
> In theory we could probably support much an older version for now, all the way back to 1.0.0 I suspect (released Jan 29 2020), and then update later as needed. What does @jni recommend here. I'm also fine with 1.3.3 I should add

had a 1-1 discussion with @jni, I think he'd like us to try `1.0.0`. If tests fail then we can bump to later release. We go back and support quite an old numpy. Would be nice to support oldest we can without compromising ourselves. I think here we should be pretty ok

**Comment by jni** [comment]
Yeah in summary, let's go with 1.0 if we can get away with it, and if we find that we need anything developed in the interim (e.g. for performance), then we can update.

**Comment by andy-sweet** [comment]
I think this is ready for review now, so I updated the status, PR title, and description appropriately. In particular, any reviewers should take a look at what this PR does not do and the known behaviors it breaks.

**Comment by andy-sweet** [comment]
> Yeah in summary, let's go with 1.0 if we can get away with it, and if we find that we need anything developed in the interim (e.g. for performance), then we can update.

I still went with `pandas>=1.0.0`, as I assume we don't to exactly pin the version - is that right? That means the CI is likely using a much newer version, but I used 1.0.0 specifically in my python 3.7 environment and haven't found any version specific issues with that yet.

File: napari/layers/labels/labels.py
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: Is there a little more detail that can be provided here? maybe a link where folks can learn more? i can see folks being confused reading this

**Review by sofroniewn — APPROVED** [review]
Love it! Left one minor comment about doc string, otherwise everything looks great. Let's get this in!!

File: napari/layers/labels/labels.py
Comment (jni) [inline napari/layers/labels/labels.py]: Do we want these to be public? Probably stuff like `validate_properties` was implicitly private but right now they are technically in our public API. Perhaps worth prefixing the functions and maybe even the whole module with `_`?

File: napari/layers/labels/labels.py
Comment (jni) [reply napari/layers/labels/labels.py]: Haha indeed I do find this confusing. Here is my suggested wording:

```
It is an implementation detail that this is a pandas DataFrame object. In the future,
we will target the currently-in-development DataAPIs [1] Dataframe API standard [2].
This will enable us to use alternate libraries such as xarray or cudf for performance
or for additional features.

If you need to rely on pandas API features, please coerce this to a pandas.DataFrame.

References
----------
.. [1]: https://data-apis.org
.. [2]: https://github.com/data-apis/dataframe-api
```

I *suspect* that `pandas.DataFrame(dataframe_api_obj)` will work in the future, but I'm not certain. If we want to be extra-safe, we should provide a public utility function `features_to_pandas_dataframe`, that we maintain until the Dataframe API is stable. What do you think @sofroniewn @andy-sweet?

File: napari/layers/labels/labels.py
Comment (jni) [inline napari/layers/labels/labels.py:481]: I think here we are missing an opportunity of making the *actual* dataframe index the index. I think that should be the default value in the case that the index of the dataframe is not equal to `np.arange(nrows)`. (This would not break backward-compatibility with the dict input.)

For me this is a huge bit of "magic" behaviour that we can now potentially close, and we should do it ASAP! (ie I favour deprecating the use of the `'index'` column: users should set the index on the dataframe.)

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: Omg :joy:

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: Same as comment above (obvs)

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: I'm confused, how are we replacing the *functionality* of choices here? iirc, we need separate choices because the dataset might not have all of the possible values already. e.g. if we are classifying cats, dogs, and bicycles, but have only encountered cats and dogs in the data so far, we still need to be able to select bicycles in the UI to add new points.

CC @kevinyamauchi who I think did this work originally... I remember being opposed to it/annoyed that we needed an extra attribute, but convinced in the end that we did need it... :joy:

Comment (jni) [reply napari/layers/points/points.py]: Holy shit I found the answer in the [features_from_properties definition](https://github.com/napari/napari/pull/3730/files#diff-0f1588377d587a70fc4bdd1dc19e5d0036b98029a9651f7d3a7d87815ab4f210R600-R608) and it's brilliant! Worth the price of admission (pd dependency) by itself I reckon. :smiley:

**Review by jni — APPROVED** [review]
Yup this looks awesome! Only a few minor updates needed afaic. I'm so looking forward to having this in!!!

**Comment by jni** [comment]
@andy-sweet 

> I still went with pandas>=1.0.0, as I assume we don't to exactly pin the version - is that right? That means the CI is likely using a much newer version, 

Indeed we don't pin, but CI does have one matrix entry with the minimum requirements, so we do test them. :blush: Currently it's failed with a timeout error :grimacing: but it seems unrelated. Let's see once you fix the merge conflicts with main.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py]: Agreed that these don't need to be public, maybe with the exception of `validate_features` the docstring and typing of which informs a user about what the `@features.setter` accepts as input - though we could just add that to the docstring of each public setter instead.

In general, I'm a bit confused about the scope utils modules and helper classes in napari. There are many public modules like this that mostly seem to contain implementation details and functions that we probably don't want to support as part of a long term public API.

So I'm definitely in favor of exposing fewer of these - to start I'll add an underscore to the functions I'm adding in this PR. Later (e.g. in version 0.5), we might want to consider making most of these modules technically private too.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py]: I like the wording above and the idea of adding the public utility function that we can reference for people that want to guarantee a pandas DataFrame.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py:481]: This is a really useful comment, as I mostly refactored this blind without thinking too much and trying to be safe.

I like the idea of requiring the user to set `DataFrame.index`, but my biggest concern would be that it may not be compatible with the initial version of the DataFrame protocol, as there is [a documented design decision to not support row names initially](https://data-apis.org/dataframe-protocol/latest/API.html#design-decisions), instead preferring an approach like this where the index/label/name is a column.

But at least we could drop the implementation detail of `Labels._label_index` and use `DataFrame.index` while we are using pandas DataFrame as our current implementation.

File: napari/layers/utils/layer_utils.py
Comment (andy-sweet) [inline napari/layers/utils/layer_utils.py:420]: Related to the public/private utils discussion, this and other functions here are technically public, but I'm just deleting them. Is that OK?

File: napari/layers/vectors/_tests/test_vectors.py
Comment (andy-sweet) [inline napari/layers/vectors/_tests/test_vectors.py:246]: Note that without adding an extra element, this did not fail because pandas will successfully broadcast a column of length 1 to match any length column (i.e. 1 -> 10 works, but 2 -> 10 doesn't). I think that's fine, but just noting for reference.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py]: How about the following?

```
    @property
    def features(self):
        """ Dataframe-like features table.

        It is an implementation detail that this is a `pandas.DataFrame`. In the future,
        we will target the currently-in-development Data API dataframe protocol [1].
        This will enable us to use alternate libraries such as xarray or cuDF for
        additional features without breaking existing usage of this.

        If you need to specifically rely on the pandas API, please coerce this to a
        `pandas.DataFrame` using `features_to_pandas_dataframe`.

        References
        ----------
        .. [1]: https://data-apis.org/dataframe-protocol/latest/API.html
        """
```

My only concern is that this does not tell the user what they can do with this (other than pass it to `features_to_pandas_dataframe`), only what they might be able to do in the future.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py:481]: We decided to leave this alone for now, but will probably follow-up with a small change here.

File: napari/layers/utils/layer_utils.py
Comment (andy-sweet) [reply napari/layers/utils/layer_utils.py:420]: These are back, so no problem anymore.

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py]: We decided that this docstring should suffice for now. We'll try to improve as needed and as people start to use features.

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py]: @jni , @tlambert03 : I tried a few things, but I landed on this. I realized that `Layer._get_state` is called in `_write_single_layer_with_plugins` *and* `Layer.as_layer_data_tuple`.

This change means I only need this temporary workaround in one place. All these `_get_state` methods are a little ugly, but they are private right now. My guess is that this isn't the last piece of state that we deprecate, so we might want to think about how to handle this more generally. One simple change is to make the abstract method that subclasses implement something like `_get_layer_specific_state` and make the base `Layer` just implement `_get_state` by adding the base state, appending the layer specific state, then adding any deprecated state.

Comment (andy-sweet) [reply napari/layers/base/base.py]: Actually, this won't work... see the `test_layer_save.py` failure. I'll think a little more on this, or feel free to add ideas.

Comment (andy-sweet) [reply napari/layers/base/base.py]: I'm starting to think the simplest thing to do is just leave `properties` and `property_choices` in the return value of `_get_state`, as well as the new `features`.

When we deprecate `properties` and `property_choices` in a follow-up PR (but continue to include them in `_get_state`), this will cause `test_layer_save.py` and `test_serialize.py` to fail due to at least one deprecation warning, but maybe ignoring or guarding against those warnings in the tests is OK?

Another option is to add a `include_deprecated` parameter to `as_layer_data_tuple` and `_get_state` that defaults to False, but is True when that state is exposed to plugins. That adds a little extra implementation to the Layers, but not much. I mostly just don't like adding a parameter that could be a temporary workaround, but maybe if Layers as EventedModels is close then this problem will go away?

@sofroniewn : could also be useful to get your perspective here or any other ideas

Comment (jni) [reply napari/layers/base/base.py]: > Another option is to add a include_deprecated parameter to as_layer_data_tuple and _get_state that defaults to False, but is True when that state is exposed to plugins. That adds a little extra implementation to the Layers, but not much. I mostly just don't like adding a parameter that could be a temporary workaround, but maybe if Layers as EventedModels is close then this problem will go away?

This would be the option I would choose, but I'm also ok with "just" leaving both items on the dict for now. Ultimately I don't think this is *hugely* important to get exactly right because a code search of `napari_get_reader properties` finds zero uses in the wild (only our own built-ins). So I would go with what seems easiest.

File: napari/layers/labels/labels.py
Comment (sofroniewn) [reply napari/layers/labels/labels.py]: > In general, I'm a bit confused about the scope utils modules and helper classes in napari. There are many public modules like this that mostly seem to contain implementation details and functions that we probably don't want to support as part of a long term public API.
>
> So I'm definitely in favor of exposing fewer of these - to start I'll add an underscore to the functions I'm adding in this PR. Later (e.g. in version 0.5), we might want to consider making most of these modules technically private too.

Yeah, i think a lot of these are public as unintentional oversights - i'd love to start pulling these back with `_`. Possibly have some deprecation warnings and then make private in `0.5`

File: napari/layers/base/base.py
Comment (sofroniewn) [reply napari/layers/base/base.py]: Yeah. I think it's fine to break a little here. Once we make the layers pydantic models we'll drop `_get_state` in favour of pydantics to_dict and to_json (or similarly named) methods

Comment (andy-sweet) [reply napari/layers/base/base.py]: I reverted to just leaving all of `features`, `properties`, and `property_choices` in the state dict. Will revisit this when deprecating `properties`.

**Comment by sofroniewn** [comment]
All tests are now passing with `pandas>=1.1.0` 🥳 

i will merge now!!

**Comment by jni** [comment]
:tada:

**Pull request merged**
